Initial draft of user interactions (#1)
aseriousbiz/blog: _posts/2021/2021-06-08-user-interface-support.md

---
title: "Support User Interactions With Buttons"
description: "Abbot skills can now support user interactions through user interface elements. In the latest release, Abbot skills may include buttons in a response that the user can click (Teams and Slack only)."
tags: [abbot, ui, buttons]
excerpt_image: 
author:
    avatar: https://2.gravatar.com/avatar/cdf546b601bf29a7eb4ca777544d11cd?s=160
    name: haacked
---

> Abbot skills get a lot done despite being text based. But sometimes a human just wants to click a button. With the latest release of Abbot (`0.15.0-beta`), Abbot supports user interface elements such as buttons. This enables skill authors to create approval workflows, simple games, etc. within chat.

Reviewer: ```suggestion
Abbot skills get a lot done despite being text-based; but sometimes a human just wants to click a button. With the latest release of Abbot (`0.15.0-beta`), Abbot supports user interface elements such as buttons. This enables skill authors to create approval workflows, simple games, and more within chat.
```

> Be sure to read the [documentation for the feature](https://docs.ab.bot/guides/ui-elements/?tabs=tabid-cs).

Reviewer: This feels like it could be pushed down to the end of the blog post.

> In this blog post, we build on the documentation to build a fun little Bot or Not skill. The skill is a game where the user is presented with an image and clicks a button to indicate whether the image is a bot or not.

Reviewer: ```suggestion
In this blog post, we build on the documentation to build a fun little "Bot or Not" skill. The skill is a game where the user is presented with an image and clicks a button to indicate whether the image is a bot or not.
```

        new("Robby the Robot", "https://user-images.githubusercontent.com/19977/119185116-375af500-ba2b-11eb-8e5b-f46ff71b93b4.png", true)
    };
}

async Task AskBotOrNotAsync(IReadOnlyList<Candidate> candidates) {
    var randomCandidate = Bot.Utilities.GetRandomElement(candidates);
    var isBot = randomCandidate.IsBot;
    
    var yesAnswer = "yes " + (isBot ? "correct" : "wrong");
    var noAnswer = "no " + (!isBot ? "correct" : "wrong");
    
    await Bot.ReplyWithButtonsAsync(
        "What do you think?",
        new Button[] {
            new("Yes, that's a bot", yesAnswer),
            new("No, not a bot.", noAnswer)
        },
        "Bot or not?",
        new Uri(randomCandidate.Url),
        randomCandidate.Title,
        "#800080");
}

public class Candidate {
    public Candidate(string title, string url, bool isBot) {
        Title = title;
        Url = url;
        IsBot = isBot;
    }
    public string Title { get; set; }
    public string Url { get; set; }
    public bool IsBot { get;set; }
}
```

### Bot or Not JavaScript

```js

> module.exports.BotOrNotJs = (async () => { // We modularize your code in order to run it.

Reviewer: ```suggestion

```

      {title: "Will Smith", url: "https://user-images.githubusercontent.com/19977/119184726-9cfab180-ba2a-11eb-9c04-7eeca7ea69c0.png", isBot: false},
      {title: "Tree", url: "https://user-images.githubusercontent.com/19977/119185238-66716680-ba2b-11eb-93a7-33e452961ffd.png", isBot: false},
      {title: "Princess Leia", url: "https://user-images.githubusercontent.com/19977/119184920-db906c00-ba2a-11eb-8523-109fe87556be.png", isBot: false},
      {title: "Robby the Robot", url: "https://user-images.githubusercontent.com/19977/119185116-375af500-ba2b-11eb-8e5b-f46ff71b93b4.png", isBot: true}
    ];
  }

  async function askBotOrNot(candidates) {
      const randomCandidate = _.sample(candidates);
      const isBot = randomCandidate.isBot;

      const yesAnswer = "yes " + (isBot ? "correct" : "wrong");
      const noAnswer = "no " + (!isBot ? "correct" : "wrong");

      await bot.replyWithButtons(
        "What do you think?",
        [new Button("Yes, that's a bot", yesAnswer), new Button("No, not a bot.", noAnswer)],
        "Bot or not?",
        randomCandidate.url,
        randomCandidate.title,
        "#800080");
    
  }

> })(); // Thanks for building with Abbot!

Reviewer: ```suggestion
```

Suggest removing the function wrapper for JS examples (same as the other docs)

    Candidate("Tree", "https://user-images.githubusercontent.com/19977/119185238-66716680-ba2b-11eb-93a7-33e452961ffd.png", False),
    Candidate("Princess Leia", "https://user-images.githubusercontent.com/19977/119184920-db906c00-ba2a-11eb-8523-109fe87556be.png", False),
    Candidate("Robby the Robot", "https://user-images.githubusercontent.com/19977/119185116-375af500-ba2b-11eb-8e5b-f46ff71b93b4.png", True)
  ]

def ask_bot_or_not(candidates):
    random_candidate = random.choice(candidates)
    is_bot = random_candidate.is_bot;

    yes_answer = "yes " + ("correct" if is_bot else "wrong")
    no_answer = "no " + ("correct" if not is_bot else "wrong")

    bot.reply_with_buttons("What do you think?",
      [Button("Yes, that's a bot", yes_answer), Button("No, not a bot.", no_answer)],
      "Bot or not?",
      random_candidate.url,
      random_candidate.title,
      "#800080")

candidates = get_candidates();

if not bot.is_interaction:
    ask_bot_or_not(candidates)
else:
    answer = bot.tokenized_arguments[0]
    correct = bot.tokenized_arguments[1]

>     bot.reply('{} answered {} which is... {}! Let’s do it again!'.format(bot.from_user.get('Name'), answer, correct))

Author: Yes, because that is not a single quote but a typographic apostrophe. However, this should probably be double quotes anyways, right?
